File: 0200/dfs.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <limits>
#include <numeric>
#include <queue>
#include <set>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  int numIslands(vector<vector<char>>& grid) {
    if (grid.size() == 0) return 0;
    int ret = 0;
    for (int i = 0, m = grid.size(); i < m; i++) {
      for (int j = 0, n = grid[0].size(); j < n; j++) {
        if (grid[i][j] == '1') {
          dfs(grid, i, j);
          ret++;
        }
      }
    }
    return ret;
  }

  void dfs(vector<vector<char>>& grid, int i, int j) {
    if (i < 0 || i >= grid.size() || j < 0 || j >= grid[0].size()) return;
    if (grid[i][j] != '1') return;
    grid[i][j] = '-';
    dfs(grid, i-1, j);
    dfs(grid, i+1, j);
    dfs(grid, i, j-1);
    dfs(grid, i, j+1);
  }
};

int main() {
  Solution sol;
  vector<vector<char>> grid;

  grid = {
    {'1','1','1','1','0'},
    {'1','1','0','1','0'},
    {'1','1','0','0','0'},
    {'0','0','0','0','0'}
  };
  assert(sol.numIslands(grid) == 1);

  grid = {
    {'1','1','0','0','0'},
    {'1','1','0','0','0'},
    {'0','0','1','0','0'},
    {'0','0','0','1','1'}
  };
  assert(sol.numIslands(grid) == 3);

  grid = {};
  assert(sol.numIslands(grid) == 0);

  grid = {
    {'1','0','0','1','0'},
  };
  assert(sol.numIslands(grid) == 2);
}

File: 0200/dfs.rs
struct Solution;

impl Solution {
    pub fn num_islands(mut grid: Vec<Vec<char>>) -> i32 {
        let m: usize = grid.len();
        let mut ret: i32 = 0;
        for i in 0..m {
            let n: usize = grid[i].len();
            for j in 0..n {
                if grid[i][j] == '1' {
                    Solution::dfs(&mut grid, i, j);
                    ret += 1;
                }
            }
        }
        return ret;
    }

    pub fn dfs(grid: &mut Vec<Vec<char>>, i: usize, j: usize) {
        let m: usize = grid.len();
        let n: usize = grid[i].len();
        if grid[i][j] != '1' { return; }
        grid[i][j] = '-';
        if i > 0 { Solution::dfs(grid, i-1, j); }
        if i < m-1 { Solution::dfs(grid, i+1, j); }
        if j > 0 { Solution::dfs(grid, i, j-1); }
        if j < n-1 { Solution::dfs(grid, i, j+1); }
    }
}

fn main() {
    let mut grid: Vec<Vec<char>>;
    grid = vec![vec!['0'; 5]; 4];
    grid[0][0] = '1';
    grid[0][1] = '1';
    grid[0][2] = '1';
    grid[0][3] = '1';
    grid[1][0] = '1';
    grid[1][1] = '1';
    grid[1][3] = '1';
    grid[2][0] = '1';
    grid[2][1] = '1';
    assert_eq!(Solution::num_islands(grid), 1);

    grid = vec![vec!['0'; 5]; 4];
    grid[0][0] = '1';
    grid[0][1] = '1';
    grid[1][0] = '1';
    grid[1][1] = '1';
    grid[2][2] = '1';
    grid[3][3] = '1';
    grid[3][3] = '1';
    assert_eq!(Solution::num_islands(grid), 3);
}
